### ISSUE TITLE
Type checking does not terminate

### ISSUE THREAD

**Issue by dm606** [agda/agda#2627]
Type checking fails to terminate on the following code:

```agda
module Bug where

data ⊤ : Set where
  tt : ⊤

data _≡_ (a : ⊤) : {B : Set} → B → Set where
  refl : a ≡ a

foo : ⊤
foo = (λ (z : ⊤) (q : bar ≡ ?) → tt) bar refl
  where
    bar : ⊤
    bar = ?
```

The type checker does terminate if the code is changed to make `bar` into a top-level declaration, use homogeneous equality, or declare `⊤` using `record` instead of `data`.

**Comment by andreasabel** [comment]
I am afraid you are indirectly defining `bar = bar` before the termination checker gets a chance to catch this loop.

**Comment by andreasabel** [comment]
Type-based termination #695 would fix this, I suppose.

**Comment by dm606** [comment]
I'm not sure I understand. How is this defining `bar = bar`? The hole in the definition of `bar` could have the solution `tt`, so surely the type checker shouldn't instantiate it with `bar`.

**Comment by andreasabel** [comment]
Ok, maybe not `bar = bar`, but `bar = ?1 bar` which has an infinite reduction sequence
`bar --> ?1 bar --> ?1 (?1 bar) --> ...`.  The derivation could go like this:
```
bar = ?0
q : bar == ?1 z
q : bar == ?1 bar
refl : bar == ?1 bar
refl : ?0 == ?1 bar
?0 := ?1 bar
bar = ?1 bar
```
You can actually see that if you put the `refl` into a shed.
```agda
foo : Unit
foo = (λ (z : Unit) (q : bar ≡ {!!}) → unit) bar {! refl !}
  where
    bar : Unit
    bar = {!!}
```
Agda terminates, and the goal buffer shows:
```
?0 : Unit
?1 : _B_16 z
?2 : bar ≡ ?1 bar
```
One you instantiate `?2` with `refl`, this produces the definitional equality `bar = ?1 bar` which can cause Agda to loop.

**Comment by UlfNorell** [comment]
Fixed since `2.5.3`.
